models/convert/cadence.go

```diff
@@ -178,3 +180,50 @@ func ParseCadenceArgument(param string) (cadence.Value, error) {
 		return nil, fmt.Errorf("unknown type for Cadence conversion (%s)", typ)
 	}
 }
+
+// ParseRosettaValue parses the input string, which should be the Rosetta value, and
+// produces the Cadence UFix64 number. The function will assume that the input string
+// has dps.FlowDecimals decimal places.
+func ParseRosettaValue(value string) (cadence.UFix64, error) {
```

Comment:
That's a really risky way to handle it. There should be no ambiguity / possibility to use this function wrong.

It should split the input into two parts using the decimal points; then it should check that the second part is smaller or equal to `dps.FlowDecimals`. If it's bigger, it should error.

If the split doesn't work, it should try parsing the first part as a number without decimals.

Reply:
The thing is, Rosetta values don't really have a decimal point.. They deal with atomic values, which are the smallest units, so the values are integers, and the number of (potential) decimals is denoted in the currency identifier.. So e.g. 1.00000001 FLOW tokens will be "100000001" with the currency identifier saying it has 8 decimals. So we can't really split on the decimal point.

In the new PR, I've made the number of decimals a function parameter. Since we're also checking that as part of our currency validation, I've made it error if the value != `dps.FlowDecimals`.

I originally went through a different route, where the number of decimals was flexible, but I didn't like how it turned out, the code felt fragile (that's how I feel about most string parsing code in scenarios like this). [Here's a link to the commit](https://github.com/optakt/flow-dps/pull/344/commits/025152a8cf10db2cb8fc940839813476981c8128).

Since the `flow-go-sdk` cadence conversion code assumes 8 decimal places, if I wanted to have a flexible number of decimals, I needed to use `cadence.NewUFix64`, which does string parsing.

I'm not too happy with any of the options. There's also the alley of parsing the number as something like an `int64` or `big.Int` and assign it directly to the `cadence.UFix64`, which is `int64`. But that also feels hackish, and if something were to change in `flow-go-sdk`, there code be issues..

Reply:
I responded to this on your new PR. I think you just need to dig in a bit deeper to understand how `UFix64` is parsed from a decimal string in Cadence. You will find that it merely puts it together into an unsigned integer, that represents exactly what we already have in the Rosetta string value, and then casts it to `cadence.UFix64`. So we can simply parse a `uint64` and cast it to get the same result.